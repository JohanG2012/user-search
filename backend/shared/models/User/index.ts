import { Document, model, Schema } from 'mongoose';

interface IName extends Document {
  title: string;
  first: string;
  last: string;
}

interface IPicture extends Document {
  large: string;
  medium: string;
  thumbnail: string;
}

interface IUser extends Document {
  name: IName;
  picture: IPicture;
  permission: boolean;
}

const userSchema = new Schema(
  {
    name: {
      title: { type: String, required: true },
      first: { type: String, required: true },
      last: { type: String, required: true },
    },
    picture: {
      large: { type: String, required: true },
      medium: { type: String, required: true },
      thumbnail: { type: String, required: true },
    },
    permission: { type: Boolean },
  },
  {
    versionKey: false,
  },
);

export default model<IUser>('User', userSchema);
