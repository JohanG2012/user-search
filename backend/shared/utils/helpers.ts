export const stripObjProps = (inputArr: object[], stripArr: string[]): object[] => {
  return inputArr.map(obj => {
    const newObj = { ...obj };
    stripArr.forEach(str => {
      delete newObj[str];
    });
    return newObj;
  });
};
