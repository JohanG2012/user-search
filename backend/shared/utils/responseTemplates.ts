import { Context } from '@azure/functions';
import { HttpStatusCode } from '../enums/HttpStatusCode';
import { IHttpResponse } from '../interfaces/IHttpResponse';
import { IBody } from '../interfaces/IBody';

export const DATABASE_CONNECTION_ERROR = (): IHttpResponse => {
  const body: IBody = {
    status: 'error',
    message: 'Unable to reach database',
    errorCode: 2001,
  };

  const response: IHttpResponse = {
    status: HttpStatusCode.ServiceUnavailable,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const VALIDATION_ERROR = () => {
  // TODO: Implement, use code 2002
};

export const INTERNAL_ERROR = (e: { message: string }, context: Context): IHttpResponse => {
  context.log(e.message); // Could be set to print to log files instead
  const body: IBody = {
    status: 'error',
    message:
      'Oops! Something unknown happend. Please open an issues with intstructions on how to reproduce this response.',
    errorCode: 2003,
  };
  const response: IHttpResponse = {
    status: HttpStatusCode.serverError,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const NOT_FOUND = (): IHttpResponse => {
  const body: IBody = {
    status: 'failed',
    message: 'Could not find requested resource.',
  };

  const response: IHttpResponse = {
    status: HttpStatusCode.NotFound,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const CREATED = (): IHttpResponse => {
  const body: IBody = {
    status: 'success',
    message: 'Successfully created',
  };

  const response: IHttpResponse = {
    status: HttpStatusCode.Created,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const OK = (data: any, next?: string): IHttpResponse => {
  const body: IBody = {
    status: 'success',
    data,
  };

  if (next) {
    body.next = next;
  }

  const response: IHttpResponse = {
    status: HttpStatusCode.Ok,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const DELETED = (): IHttpResponse => ({
  status: HttpStatusCode.NoContent,
});

export const UPDATED = (): IHttpResponse => ({
  status: HttpStatusCode.NoContent,
});
