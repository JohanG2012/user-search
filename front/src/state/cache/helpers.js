export const generateSearchCache = (state, action) => {
  const newCache = {};
  newCache[action.payload.queryString] = {
    data: action.payload.data,
    next: action.payload.next,
  };
  const search = {
    ...state.search,
    ...newCache,
  };

  let entries = Object.entries(search);
  if (entries.length > 50) {
    entries = entries.slice(entries.length - 50);
  }

  const newSearch = {};

  entries.forEach(([key, value]) => {
    newSearch[key] = value;
  });

  return {
    search: {
      ...newSearch,
    },
  };
};

export default {
  generateSearchCache,
};
