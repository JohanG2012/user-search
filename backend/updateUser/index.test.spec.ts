/* tslint:disable no-var-requires */
const mongoose = require('mongoose');
/* tslint:enable no-var-requires */
import { MongoMemoryServer } from 'mongodb-memory-server';
import httpFunction from './index';
import User from '../shared/models/User';
import * as testData from '../data/testData.json'; // 100 user documents
import { fakeObjectId } from '../shared/utils/fakes';
import { stripObjProps } from '../shared/utils/helpers';
mongoose.Promise = global.Promise;

let mongoServer;

beforeAll(async () => {
  mongoServer = new MongoMemoryServer();
  const mongoUri = await mongoServer.getConnectionString();
  await mongoose.connect(mongoUri, err => {
    if (err) {
      console.error(err);
    }
  });

  // Promise fix for broken insertMany
  await new Promise((resolve, reject) => {
    User.insertMany(testData, err => {
      if (err) {
        reject(err);
      }
      resolve();
    });
  });
});

afterAll(async () => {
  await mongoose.disconnect();
  await mongoServer.stop();
});

describe('PATCH /api/v1/user/:id', () => {
  let context: any = {};
  let req: any = {};
  beforeEach(() => {
    context = {
      res: { body: {} },
      done: jest.fn(),
      log: jest.fn(),
    };

    req = {
      params: {},
      query: {},
      body: {},
    };
  });

  describe('validation', () => {
    describe('with invalid id', () => {
      beforeEach(async () => {
        req.params.id = 'fake id';
        await httpFunction(context, req);
      });

      it('should return 400 status code', async () => {
        expect(context.res.status).toBe(400);
      });

      it('should match snapshot', () => {
        expect(context).toMatchSnapshot();
        expect(context.done).toHaveBeenCalledTimes(1);
      });
    });
    describe('with forbidden fields only', () => {
      beforeEach(async () => {
        req.params.id = fakeObjectId();
        req.body._id = 'new _id';
        await httpFunction(context, req);
      });

      it('should return 400 status code', async () => {
        expect(context.res.status).toBe(400);
      });

      it('should match snapshot', () => {
        expect(context).toMatchSnapshot();
        expect(context.done).toHaveBeenCalledTimes(1);
      });
    });
    describe('with empty body', () => {
      beforeEach(async () => {
        req.params.id = fakeObjectId();
        req.body = {};
        await httpFunction(context, req);
      });

      it('should return 400 status code', async () => {
        expect(context.res.status).toBe(400);
      });

      it('should match snapshot', () => {
        expect(context).toMatchSnapshot();
        expect(context.done).toHaveBeenCalledTimes(1);
      });
    });
    describe('with non existing _id', () => {
      beforeEach(async () => {
        req.params.id = fakeObjectId();
        req.body.permission = true;
        await httpFunction(context, req);
      });

      it('should return 400 status code', async () => {
        expect(context.res.status).toBe(404);
      });

      it('should match snapshot', () => {
        expect(context).toMatchSnapshot();
        expect(context.done).toHaveBeenCalledTimes(1);
      });
    });
  });
  describe('update an user', () => {
    let original: any;
    const newFirstName = 'updated';
    let FAKE_ID: string;
    let newUser: any;
    beforeEach(async () => {
      // MongoMemoryServer id's wont pass validation as a real mongoDB id.
      // First need create a document with fake_id.
      FAKE_ID = fakeObjectId();
      const [user]: any = await User.find().limit(1);
      newUser = { ...user._doc, _id: FAKE_ID };
      await new User(newUser).save();

      // Update
      req.params.id = FAKE_ID;
      req.body = { name: { first: newFirstName } };
      await httpFunction(context, req);
    });

    it('should respond with 204 status code', () => {
      expect(context.res.status).toBe(204);
    });

    it('should match snapshot', () => {
      expect(context).toMatchSnapshot();
      expect(context.done).toHaveBeenCalledTimes(1);
    });

    it('should be able to update a item', async () => {
      const [updated] = await User.find({ _id: FAKE_ID });
      expect(updated.name.first).toBe(newFirstName);
    });

    it('should only change body data', async () => {
      const [updated]: any = await User.find({ _id: FAKE_ID });
      original = { ...newUser };
      original.name.first = newFirstName;
      const [stripedUpdated] = stripObjProps([updated._doc], ['_id']);
      const [stripedOriginal] = stripObjProps([original], ['_id']);
      expect(stripedUpdated).toMatchObject(stripedOriginal);
    });
  });
  describe('trigger server error - close down database connection', () => {
    beforeEach(async () => {
      mongoose.disconnect();
      mongoServer.stop();
      req.params.id = fakeObjectId();
      req.body = { name: { first: 'updated' } };
      await httpFunction(context, req);
    });

    it('should respond with 503 error', async () => {
      expect(context.res.status).toBe(503);
    });

    it('should match snapshot', async () => {
      expect(context).toMatchSnapshot();
      expect(context.done).toHaveBeenCalledTimes(1);
    });
  });
});
