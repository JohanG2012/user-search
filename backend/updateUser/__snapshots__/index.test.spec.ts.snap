// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PATCH /api/v1/user/:id trigger server error - close down database connection should match snapshot 1`] = `
Object {
  "done": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "log": [MockFunction] {
    "calls": Array [
      Array [
        "topology was destroyed",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "res": Object {
    "body": "{\\"status\\":\\"error\\",\\"message\\":\\"Oops! Something unknown happend. Please open an issues with intstructions on how to reproduce this response.\\",\\"errorCode\\":2003}",
    "headers": Object {
      "Content-Type": "application/json",
    },
    "status": 500,
  },
}
`;

exports[`PATCH /api/v1/user/:id update an user should match snapshot 1`] = `
Object {
  "done": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "log": [MockFunction],
  "res": Object {
    "status": 204,
  },
}
`;

exports[`PATCH /api/v1/user/:id validation with empty body should match snapshot 1`] = `
Object {
  "done": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "log": [MockFunction],
  "res": Object {
    "body": "{\\"status\\":\\"failed\\",\\"message\\":\\"\\\\\\"value\\\\\\" must contain at least one of [permission, picture.large, picture.medium, picture.thumbnail, name.title, name.first, name.last]\\",\\"errorCode\\":2002}",
    "headers": Object {
      "Content-Type": "applicaiton/json",
    },
    "status": 400,
  },
}
`;

exports[`PATCH /api/v1/user/:id validation with forbidden fields only should match snapshot 1`] = `
Object {
  "done": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "log": [MockFunction],
  "res": Object {
    "body": "{\\"status\\":\\"failed\\",\\"message\\":\\"\\\\\\"value\\\\\\" must contain at least one of [permission, picture.large, picture.medium, picture.thumbnail, name.title, name.first, name.last]\\",\\"errorCode\\":2002}",
    "headers": Object {
      "Content-Type": "applicaiton/json",
    },
    "status": 400,
  },
}
`;

exports[`PATCH /api/v1/user/:id validation with invalid id should match snapshot 1`] = `
Object {
  "done": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "log": [MockFunction],
  "res": Object {
    "body": "{\\"status\\":\\"failed\\",\\"message\\":\\"\\\\\\"value\\\\\\" must contain at least one of [permission, picture.large, picture.medium, picture.thumbnail, name.title, name.first, name.last]\\",\\"errorCode\\":2002}",
    "headers": Object {
      "Content-Type": "applicaiton/json",
    },
    "status": 400,
  },
}
`;

exports[`PATCH /api/v1/user/:id validation with non existing _id should match snapshot 1`] = `
Object {
  "done": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "log": [MockFunction],
  "res": Object {
    "body": "{\\"status\\":\\"failed\\",\\"message\\":\\"Could not find requested resource.\\"}",
    "headers": Object {
      "Content-Type": "applicaiton/json",
    },
    "status": 404,
  },
}
`;
